### PR TITLE
Headless SVG rendering

### PR DESCRIPTION
**Pull request by marcprux**
Implementation of SVG headless rendering mode for https://github.com/uwdata/vega/issues/10. This PR replaces https://github.com/uwdata/vega/pull/21

### CODE REVIEW

**Comment by marcprux** [comment]
A few more notes about this PR:
- Unlike #21, I didn't implement the part of the test cases that renders the same spec using jsdom and compares the XML output with the headless output. It was very slow and rather flaky, but I can put some work into re-implementing it if there is any interest.
- I named the headless renderer "svg-xml", mostly because it was terser than "svg-headless".
- Some of the example spec rendering test cases are disabled for now pending https://github.com/uwdata/datalib/pull/4

File: gulpfile.js [hunk 1]
```diff
@@ -2,12 +2,14 @@ var gulp = require('gulp'),
     browserify = require('browserify'),
     buffer = require('vinyl-buffer'),
     rename = require('gulp-rename'),
+    run = require('gulp-run'),
```
Comment (arvind) [inline gulpfile.js]: @mprudhom: Is this used for anything? If not, can we drop it here and from `package.json`?

**Comment by arvind** [comment]
I would prefer to keep the `svg-headless` name that we agreed upon in #21. This will keep it consistent with other headless renderers we may add moving forward (e.g., a canvas headless renderer).

File: src/core/View.js [hunk 2]
```diff
@@ -135,39 +140,47 @@ prototype.initialize = function(el) {
...
-    .append("div")
-    .attr("class", "vega")
-    .style("position", "relative")
-    .node();
-  if (v._viewport) {
-    d3.select(el)
-      .style("width",  (v._viewport[0] || w)+"px")
-      .style("height", (v._viewport[1] || h)+"px")
-      .style("overflow", "auto");
+  if (!arguments.length || el === null) {
```
Comment (arvind) [inline src/core/View.js]: @mprudhom: Do you mind explaining this change and the ones that follow to this method? Not sure I understand the need for the change as it relates to headless rendering, or why only certain parts of the method check for `el`. Do you have a use case where this is necessary (i.e., where one would want to re-initialize the visualization but, for some reason, cannot specify an element)?

**Comment by arvind** [comment]
One more question -- I see you've implemented another SVG Renderer under svg-xml. I'm curious why we weren't able to just use the existing one? Looking at Vega 1, it seemed to reuse the same renderer for both web-based and server-side/headless rendering?

**Comment by arvind** [comment]
Ah never mind, I see we introduced a new headless renderer in Vega 1.

File: gulpfile.js (hunk 1 above)
Comment (marcprux) [inline gulpfile.js]: @arvind I've removed it. I don't recall why I put it there.

**Comment by marcprux** [comment]
OK, I've changed `svg-xml` back to `svg-headless`. Let me know if you need any other fixes.

**Comment by arvind** [comment]
Thanks for the PR, Marc! I've merged it into the v2 branch and have included the `vg2png` and `vg2svg` binaries as well. 

I did introduce a change from what we'd discussed in the previous PR (#21). Instead of merging the headless and browser `Views`, I broke them back out into separate classes (with our new OO architecture, we can do this in a more elegant way than Vega 1). There were a few reasons I ended up preferring two separate classes:
- Headless renderers may require special methods (e.g., headless canvas requires a separate initialization and an async callback). It would be nice not to pollute the browser `View` with them. 
- It keeps the public API clean -- I really liked your implementation of enabling headless rendering if no `el` was specified, but it required users to be aware that a separate `svg-headless` renderer needed to be called (while `canvas` was correct for both browser and headless views). Although one could argue that we could maintain a clean public API with a single `View` class, a separate `HeadlessView` also allows us to...
- Minimize the number of `if(el)` statements. You'll notice, however, that I've maintained this logic that you introduced -- regardless of browser or headless, users call `vg.parse.spec` and the `ViewFactory` determines which `View` is produced. 

Thanks again for the PR (and for including test cases!). Let me know if you have any questions or notice any issues.

**Comment by marcprux** [comment]
Arvind-

Great, I'm thrilled to see it get merged in! The breakdown into separate classes is fine with me.

```
-Marc
```

**Pull request merged**